apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/ClientPoolTests.cs

```diff
@@ -0,0 +1,171 @@
+            var mockClientFactory = SetupMockClientFactory(firstReturnedClient);
+
+            var testPool = new ClientPool(mockClientFactory.Object);
+            var clientAcquired = await testPool.AcquireClientAsync();
+            testPool.ReleaseClient(clientAcquired);
+
+            mockClientFactory.Verify(x => x.CreateClientAsync(It.IsAny<ServiceEndpoint>(), It.IsAny<ITokenAccess>()), Times.Exactly(1));
+            var s = new StringBuilder();
+            Assert.Equal(0, testPool.GetPoolInfo(s));
+            Logger.Instance.Info(s.ToString());
+        }
+
+        [Fact]
+        public async Task SequentialCreatesOnlyUseOneChannel()
+        {
+            //SpannerClients if created sequentially should only use a single channel.
+            //This is to faciliate good session pool hits (which is per channel).  If we always
+            //round robin'd clients, then we would get hard cache misses until we populated all
+            //of the caches per channel.
+            //Since channels are lazily created, this also saves resources for scenarios like rich clients
+            //that may only need a single session for multiple sequential workloads.
+            //Our default setting for # channels = 4 needs to work well for all scenarios.
+            var firstReturnedClient = new Mock<SpannerClient>();
+            var mockClientFactory = SetupMockClientFactory(firstReturnedClient);
+
+            var testPool = new ClientPool(mockClientFactory.Object);
+            var expectedClient = firstReturnedClient.Object;
+
+            Assert.Same(expectedClient, await testPool.AcquireClientAsync());
+            testPool.ReleaseClient(expectedClient);
+
+            Assert.Equal(expectedClient, await testPool.AcquireClientAsync());
+            testPool.ReleaseClient(expectedClient);
+
+            Assert.Equal(expectedClient, await testPool.AcquireClientAsync());
+            testPool.ReleaseClient(expectedClient);
+
+            var s = new StringBuilder();
+            Assert.Equal(0, testPool.GetPoolInfo(s));
+            Logger.Instance.Info(s.ToString());
+            mockClientFactory.Verify(x => x.CreateClientAsync(It.IsAny<ServiceEndpoint>(), It.IsAny<ITokenAccess>()), Times.Exactly(1));
+        }
+
+        [Fact]
+        public async Task RoundRobin()
+        {
+            var firstReturnedClient = new Mock<SpannerClient>();
+            var mockClientFactory = SetupMockClientFactory(firstReturnedClient);
+
+            var testPool = new ClientPool(mockClientFactory.Object);
+
+            List<SpannerClient> clientList = new List<SpannerClient>();
+            for (var i = 0; i < SpannerOptions.Instance.MaximumGrpcChannels; i++)
+            {
+                var newClient = await testPool.AcquireClientAsync();
+                foreach (var existing in clientList)
+                {
+                    Assert.NotSame(existing, newClient);
+                }
```

Comment:
Yes it will, but if that's a problem, then line 155 below will be as well: `GroupBy` groups keys using the default equality comparer, which uses `Equals`. The clients don't appear to be overriding `Equals`, so I assume they should both be fine.